Server/JokesOnYou.Web.Api/Controllers/UserController.cs

```diff
@@ -1,48 +1,143 @@
         {
-            throw new NotImplementedException();
+            return Ok(_userService.GetAll());
         }
 
+        //idk if its from body or Http get thing
         [HttpGet("{id}")]
-        public async Task<ActionResult> GetUser(string id)
+        public  ActionResult<User> GetUserById(string id)
         {
-            throw new NotImplementedException();
-        }
+            var user =  _userService.GetUserById(id);
 
-        [HttpPatch]
-        public async Task<ActionResult> UpdateUser(UserUpdateDTO userUpdateDTO)
+            if (user != null)
+            {
+                return user;
+            }
+            _logger.LogInformation($"User not found; id: {id}");
+            return NotFound();
+            
+        }
+        [HttpPut("{id}")]
+        public async Task<ActionResult> UpdateUser(string id, [FromBody]UserUpdateDTO userUpdateDTO)
         {
-            throw new NotImplementedException();
+            var user = _mapper.Map<User>(userUpdateDTO);
+
+
+            if (await _userService.UpdateUser(user))
+            {
+                return Ok();
+            }
+            _logger.LogInformation($"Something went wrong with updating user that had id {user.Id}");
+
+            return NotFound();
         }
 
+        [AllowAnonymous]
+        [HttpPost("authenticate")]
+        public async Task<ActionResult> Authenticate([FromBody] UserLoginDTO userDto)
+        {
+            User user;
+            if (string.IsNullOrWhiteSpace(userDto.Password))
+            {
+                return Unauthorized(); //or something else, idk
+            }
+
+            //maybe use a switch? should be much different
+            if (!string.IsNullOrWhiteSpace(userDto.Username))
+            {
+                user =await _userService.Authenticate(userDto.Username, userDto.Password);
+            }
+            else if (!string.IsNullOrWhiteSpace(userDto.Email))
+            {
+                user =await _userService.Authenticate(userDto.Email, userDto.Password);
+            }
+            else
+            {
+                return Unauthorized();
+            }
+
+            if (user == null)
+            {
+                _logger.LogInformation($"User / possible attacker inserted wrong password on account id: {user.Id}");
+                return Unauthorized();
+            }
+                
+
+            _tokenService.GetToken(user);
+            return Ok(new
+            {
+                Id = user.Id,
+                Username = user.UserName,
+                Email = user.Email,
+                Token = _tokenService.GetToken(user)
+        });
+        }
+        
         [HttpDelete("{id}")]
-        public async Task<ActionResult> DeleteUser(string id)
+        public ActionResult DeleteUser(string id)
         {
-            throw new NotImplementedException();
+            var user = _userService.GetUserById(id);
+            if (user == null)
```

Comment:
all the user == null will be handled by the service since we want controllers to be clean and its considered business logic.

Reply:
What do you mean by "will be handled by the service", by checking if its null we can prevent a crash or an exception to be thrown

Reply:
the controllers are not responsible for business logic. The service will have this code. The controller is just like a guy at the post office, organizes where the mail should go next

Reply:
Makes sense